src/hub/action_request.ts

```diff
@@ -144,27 +145,47 @@ export class ActionRequest {
    * Streaming generally occurs only if Looker sends the data in a streaming fashion via a push url,
    * however it will also wrap non-streaming attachment data so that actions only need a single implementation.
    *
-   * @returns The return value of the `callback` function. This is useful for returning
-   * a promise from the `callback` function.
+   * @returns A promise returning the same value as the callback's return value.
+   * This promise will resolve after the stream has completed and the callback's promise
+   * has also resolved.
    * @param callback A function will be caled with a Node.js `Readable` object.
    * The readable object represents the streaming data.
    */
-  stream<T>(callback: (readable: Readable) => T): T {
-    const url = this.scheduledPlan && this.scheduledPlan.downloadUrl
+  async stream<T>(callback: (readable: Readable) => Promise<T>): Promise<T> {
+
     const stream = new PassThrough()
-    const returnVal = callback(stream)
-    if (url) {
-      httpRequest.get(url).pipe(stream)
-    } else {
-      if (this.attachment && this.attachment.dataBuffer) {
-        stream.end(this.attachment.dataBuffer)
+    const returnPromise = callback(stream)
+
+    const url = this.scheduledPlan && this.scheduledPlan.downloadUrl
+
+    const streamPromise = new Promise<void>((resolve, reject) => {
+      if (url) {
+        httpRequest
+          .get(url)
+          .on("error", (err) => {
+            winston.error(`Request stream error: ${err}\n${err.stack}`)
```

Comment:
do you want to keep this logging in?

Reply:
Yes, it's useful to have these errors in the action hub logs.